README.md
# validation-predicates

# Why
Functions that take some of the pain out of validating input.  Turns validation syntax into readable, easily-combinable functions.  This makes you more likely to do it.  

There are other great libraries out there that do similar things:
* [Generic validators](https://www.npmjs.com/package/validation-utils)    
* [String validation](https://github.com/validatorjs/validator.js)  
 
Please use them if you're looking for a different take on the same subject.  

# Change log  
[Link to changes](https://github.com/tmurphree/validation-predicates/blob/master/CHANGELOG.md)  

# Installation

npm install --save @tmurphree/validation-predicates  

# Usage  

``` js
const { 
  isNumberGreaterThan,
  isPopulatedArray,
  isString,  
} = require('@tmurphree/validation-predicates');

// the old way: not as readable
// const firstIsValid = function firstIsValid(x) {
//   return typeof x === 'number' &&
//     x > 5;
// };
//
// const secondIsValid = function secondIsValid(x) {
//   return Array.isArray(x) &&
//      x.length > 0 &&
//      x.every((el) => typeof el === 'string') &&
//      x.every((el) => el.length > 2);
// };

// the same tests using validation-predicates 
const firstIsValid = function firstIsValid(x) {
  return isNumberGreaterThan(x, 5);
};

const secondIsValid = function secondIsValid(x) {
  return isPopulatedArray(x) &&
    x.every(isString) &&
    // mix and match with predicates not in the package
    x.every((el) => el.length > 2);
};

const someFunction = function someFunction(first, second) {
  if (!(firstIsValid(first))) {
    throw new Error('Expected first to be a number > 5.');
  }

  if (!(secondIsValid(second))) {
    throw new Error('Expected second to be an array of strings longer than 2 characters.');
  }

  // amazing code goes here
};

someFunction(42, ['some','array']);
```

## Strict mode  
Some functions have options.  Strict mode is a convenient way to set some of the options.  **You can always, in both modes, set the options manually.**

Without strict mode:  
``` js
const { isDate, isObjectLike } = require('@tmurphree/validation-predicates');

// isDate, which has no options, is not changed

// isObjectLike(a, b) is called with { checkType: false } by default
// but you can still call isObjectLike(a, b, { checkType: true })
```

With strict mode:  
``` js
const { isDate, isObjectLike } = require('@tmurphree/validation-predicates').strict;

// isDate, which has no options, is not changed

// isObjectLike(a, b) is called with { checkType: true } by default
// but you can still call isObjectLike(a, b, { checkType: false })
```


# Functions  
* All functions return `boolean`.  
* Parameters not in brackets are required.  
* Parameters in brackets are optional.  
* 'x' in the parameters list is the thing you want to test.  

|Function|Summary|  
|---|---|  
|isArray(x)|x is an array|  
|isBigInt(x)|x is a BigInt|  
|isBoolean(x)|x is a boolean|  
|isDate(x)|x is a Date|  
|isDateAfter(x, someDate)|alias for isDateGreaterThan|  
|isDateBefore(x, someDate)|alias for isDateLessThan|  
|isDateGreaterThan(x, someDate)|x is a Date greater than (after) someDate|  
|isDateLessThan(x, someDate)|x is a Date less than (before) someDate|  
|isFloat(x)|x is a number (see isNumber) and has a nonzero decimal. e.g. 5.0 is NOT a float, but 5.01 is.|  
|isFunction(x)|x is a function|  
|isInteger(x)|x is a number (see isNumber) and has a zero-value decimal. e.g. 5 and 5.0 are both integers, but 5.01 is NOT.|  
|isIsoDateTimeString(x)|x is a string and matches a subset of the ISO 8601 date time string format.<br><br>Checks for these patterns (note that the plus can be plus or minus and that 'milliseconds' is 'one or more digits'):<br>* YYYY-MM-DDThh:mm:ssZ<br>* YYYY-MM-DDThh:mm:ss.millisecondsZ<br>* YYYY-MM-DDThh:mm:ss+hh:mm<br>* YYYY-MM-DDThh:mm:ss.milliseconds+hh:mm|    
|isNull(x)|x is null|  
|isNullOrUndefined(x)|x is null or undefined|  
|isNumber(x)|x is a number and is not NaN|  
|isNumberGreaterThan(x, anotherNumber)|x is a number and is greater than anotherNumber|  
|isNumberLessThan(x, anotherNumber)|x is a number and is less than anotherNumber|  
|isObject(x)|x is an object per the definition in 'typeof' EXCEPT that it returns false for null.  Basically, this means it isn't another known type of primitive and is not null.  See [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof).|  
|isObjectLike(x, template, [options={ allowExtraProps: false, checkType: false, debug: false }])|x is an object and has exactly the same properties as template.  x does not have properties not in template.  Does not check symbols.<br><br>**Strict mode changes the default to { checkType: true }.  Other defaults remain the same.**<br><br>Options: <br> * `allowExtraProps` Return true if the object to test has properties not in template and other tests pass.<br>* `checkType`  If true, checks that the properties are the same type **BUT** this is rudimentary checking (with typeof).<br> * `debug` If true, prints debug information to the console.|  
|isObjectWithExpectedProps(x, arrayOfStrings)|x is an object and every string in arrayOfStrings is a property of x.  x may have properties not in arrayOfStrings.|  
|isPopulatedArray(x)|x is an array with at least one element|  
|isPopulatedObject(x)|x is an object with at least one property or Symbol|
|isPopulatedString(x)|x is a string with at least one character|  
|isString(x)|x is a string|  
|isSymbol(x)|x is a [symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol)|  
|isUndefined(x)|x is undefined|  
|isZeroLength(x)|x has a length property that === 0|  


CHANGELOG.md
# 2.0.4  
Update dev dependencies to fix acorn vulnerability.  

# 2.0.3  
Set isObjectLike options without clobbering defaults  

# 2.0.2  
Add allowExtraProps option to isObjectLike  

# 2.0.1  
Streamline package  

# 2.0.0  
## Breaking changes   
* Change the default behavior of isObjectLike to use { checkType: false }  
* Replace isNotNullOrUndefined with isNullOrUndefined  
* Replace isNotZeroLength with isZeroLength  

## Other changes  
* Add strict mode  
* Add isFloat, isInteger, and isIsoDateTimeString


# 1.2.0  
* Add option to disable type checking on isObjectLike  
* Add isPopulatedObject  
* Add isDate, isDateGreaterThan, and isDateLessThan  

# 1.1.1  
* Add debug option to isObjectLike  

# 1.1  
* Add isObjectLike  
* Return false from isObject if test data is null or an array  

spec/index.spec.js
/* eslint no-undef:"off" */

const {
  isDate,
  isDateAfter,
  isDateBefore,
  isDateGreaterThan,
  isDateLessThan,
  isInteger,
  isIsoDateTimeString,
  isFloat,
  isNumber,
  isObject,
  isObjectLike,
  isObjectWithExpectedProps,
  isPopulatedArray,
  isPopulatedObject,
  isPopulatedString,
  isZeroLength,
} = require('../index');
// #region jasmine setup
const origTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL;

const revertJasmineTimeout = function revertJasmineTimeout() {
  jasmine.DEFAULT_TIMEOUT_INTERVAL = origTimeout;
};

const setJasmineTimeout = function setJasmineTimeout(milliseconds) {
  jasmine.DEFAULT_TIMEOUT_INTERVAL = milliseconds;
};

// you can set more options than are shown here: see https://jasmine.github.io/api/edge/Reporter.html
// tutorial: https://jasmine.github.io/tutorials/custom_reporter
const myReporter = {
  jasmineStarted: function jasmineStarted(suiteInfo, done) {
    // optional setup goes here
    setJasmineTimeout(10000);
    done();
  },
  jasmineDone: function jasmineDone(suiteInfo, done) {
    console.log(`Tests ended ${new Date().toLocaleString()}`);
    revertJasmineTimeout();
    done();
  },
};

jasmine.getEnv().addReporter(myReporter);
// #endregion jasmine setup

describe('isDate', () => {
  it('tests for Dates', () => {
    expect(isDate(1)).toBe(false);
    expect(isDate(new Date())).toBe(true);
  });
});

describe('isDateGreaterThan', () => {
  it('tests for Dates after another date', () => {
    const marchFirst2020 = new Date(2020, 2, 1, 0, 0, 0, 0);
    const marchFifteenth2020 = new Date(2020, 2, 15);

    expect(isDateGreaterThan('not a date', marchFifteenth2020)).toBe(false);
    expect(isDateGreaterThan(marchFirst2020, marchFifteenth2020)).toBe(false);
    expect(isDateGreaterThan(marchFifteenth2020, marchFirst2020)).toBe(true);
    // equal
    expect(isDateGreaterThan(marchFirst2020, marchFirst2020)).toBe(false);
  });

  it('has an alias', () => {
    expect(isDateAfter).toEqual(isDateGreaterThan);
  });
});

describe('isDateLessThan', () => {
  it('tests for Dates after another date', () => {
    const marchFirst2020 = new Date(2020, 2, 1, 0, 0, 0, 0);
    const marchFifteenth2020 = new Date(2020, 2, 15);

    expect(isDateLessThan('not a date', marchFifteenth2020)).toBe(false);
    expect(isDateLessThan(marchFirst2020, marchFifteenth2020)).toBe(true);
    expect(isDateLessThan(marchFifteenth2020, marchFirst2020)).toBe(false);
    // equal
    expect(isDateLessThan(marchFirst2020, marchFirst2020)).toBe(false);
  });

  it('has an alias', () => {
    expect(isDateBefore).toEqual(isDateLessThan);
  });
});

describe('isFloat', () => {
  it('tests for floats', () => {
    expect(isFloat('asdf')).toBeFalse();
    expect(isFloat(12)).toBeFalse();
    expect(isFloat(5.0)).toBeFalse();
    expect(isFloat(12.32)).toBeTrue();
  });
});

describe('isInteger', () => {
  it('tests for integers', () => {
    expect(isInteger('asdf')).toBeFalse();
    expect(isInteger(12.32)).toBeFalse();
    expect(isInteger(12)).toBeTrue();
    expect(isInteger(5.0)).toBeTrue();
  });
});

describe('isIsoDateTimeString', () => {
  it('tests for strings', () => {
    expect(isIsoDateTimeString(12)).toBeFalse();
    expect(isIsoDateTimeString(new Date())).toBeFalse();
  });

  it('tests for the +/- hh:mm format', () => {
    // error path
    const monthTensMoreThan1 = '2000-31-01T01:01:01-10:00';
    const dayTensMoreThan3 = '2000-01-41T01:01:01-10:00';
    const hourTensMoreThan2 = '2000-01-01T31:01:01-10:00';
    const minuteTensMoreThan5 = '2000-01-01T01:71:01-10:00';
    const secondTensMoreThan5 = '2000-01-01T01:01:61-10:00';
    const hourOffsetTensMoreThan2 = '2000-01-01T01:01:01-30:00';
    const minuteOffsetTensMoreThan5 = '2000-01-01T01:01:01-10:90';

    // happy path
    const noMilliseconds = '2000-01-01T01:01:01-10:00';
    const withMilliseconds = '2000-01-01T01:01:01.234-10:00';
    const usesAplusSign = '2000-01-01T01:01:01+09:00';

    expect(isIsoDateTimeString(monthTensMoreThan1)).toBeFalse();
    expect(isIsoDateTimeString(dayTensMoreThan3)).toBeFalse();
    expect(isIsoDateTimeString(hourTensMoreThan2)).toBeFalse();
    expect(isIsoDateTimeString(minuteTensMoreThan5)).toBeFalse();
    expect(isIsoDateTimeString(secondTensMoreThan5)).toBeFalse();
    expect(isIsoDateTimeString(hourOffsetTensMoreThan2)).toBeFalse();
    expect(isIsoDateTimeString(minuteOffsetTensMoreThan5)).toBeFalse();

    expect(isIsoDateTimeString(noMilliseconds)).toBeTrue();
    expect(isIsoDateTimeString(withMilliseconds)).toBeTrue();
    expect(isIsoDateTimeString(usesAplusSign)).toBeTrue();
  });

  it('tests for the Z format', () => {
    // error path
    const monthTensMoreThan1 = '2000-31-01T01:01:01Z';
    const dayTensMoreThan3 = '2000-01-41T01:01:01Z';
    const hourTensMoreThan2 = '2000-01-01T31:01:01Z';
    const minuteTensMoreThan5 = '2000-01-01T01:71:01Z';
    const secondTensMoreThan5 = '2000-01-01T01:01:61Z';
    const lowerCaseZ = '2000-01-01T01:01:01z';

    // happy path
    const noMilliseconds = '2000-01-01T01:01:01Z';
    const withMilliseconds = '2000-01-01T01:01:01.234Z';

    expect(isIsoDateTimeString(monthTensMoreThan1)).toBeFalse();
    expect(isIsoDateTimeString(dayTensMoreThan3)).toBeFalse();
    expect(isIsoDateTimeString(hourTensMoreThan2)).toBeFalse();
    expect(isIsoDateTimeString(minuteTensMoreThan5)).toBeFalse();
    expect(isIsoDateTimeString(secondTensMoreThan5)).toBeFalse();
    expect(isIsoDateTimeString(lowerCaseZ)).toBeFalse();

    expect(isIsoDateTimeString(noMilliseconds)).toBeTrue();
    expect(isIsoDateTimeString(withMilliseconds)).toBeTrue();
  });
});

describe('isNumber', () => {
  it('returns false for NaN', () => {
    expect(isNumber(NaN)).toBeFalse();
  });

  it('returns false on non-number values', () => {
    expect(isNumber('asdfasdf')).toBeFalse();
    expect(isNumber({})).toBeFalse();
  });

  it('returns true for number values, including Infinity', () => {
    expect(isNumber(12)).toBeTrue();
    expect(isNumber(12.999)).toBeTrue();
    expect(isNumber(1 / 3)).toBeTrue();
    expect(isNumber(Infinity)).toBeTrue();
    expect(isNumber(-Infinity)).toBeTrue();
  });
});

describe('isObject', () => {
  it('returns false for other known primitives', () => {
    expect(isObject(123)).toBeFalse();
    expect(isObject('string')).toBeFalse();
    // type is symbol
    expect(isObject(Symbol('s'))).toBeFalse();
  });

  it('returns true for objects', () => {
    expect(isObject({})).toBeTrue();
    expect(isObject(['an array', 'does not have its own primitive type'])).toBeTrue();
    expect(isObject(new Date())).toBeTrue();
  });
});

describe('isObjectLike', () => {
  const referenceObject = {
    age: 23,
    name: 'Bob',
    speak: (x) => console.log(x),
  };

  const makeTestObject = function makeTestObject(age = 3, name = 'Charlie') {
    const speak = (x) => console.log(x);
    return { age, name, speak };
  };

  it('throws if referenceObject is not an object', () => {
    expect(() => isObjectLike({ tested: false }, 12)).toThrow();
    expect(() => isObjectLike({ tested: false }, 'no strings, either')).toThrow();
    expect(() => isObjectLike({ tested: false }, null)).toThrow();
  });

  it('returns false if the "object" is not an object', () => {
    expect(isObjectLike(12, referenceObject)).toBe(false);
    expect(isObjectLike(null, referenceObject)).toBe(false);
  });

  it('returns true if the props are there (with type checking off by default)', () => {
    expect(isObjectLike(makeTestObject('no type checking', 'Alice'), referenceObject)).toBe(true);
    expect(isObjectLike(makeTestObject(99, 'Xavier'), referenceObject)).toBe(true);
  });

  it('returns false if the props are not there', () => {
    expect(isObjectLike({ age: 12, name: 'speak is missing' }, referenceObject)).toBe(false);
    expect(isObjectLike(['this is', 'a totally different object'], referenceObject)).toBe(false);
    expect(isObjectLike(['this is', 'a totally different object'], referenceObject, { checkType: false }))
      .toBe(false);
  });

  it('returns false if there are more props in the testObject than are in the referenceObject', () => {
    const itHasMoreProps = makeTestObject(72, 'Jans');
    itHasMoreProps.someAdditionalProp = true;
    expect(isObjectLike(itHasMoreProps, referenceObject)).toBe(false);
  });

  it('optionally checks types', () => {
    // wrong type, no type checking
    expect(isObjectLike(
      makeTestObject('oops', 'wrong type'),
      referenceObject,
      { checkType: false },
    )).toBe(true);

    // wrong type, with type checking
    expect(isObjectLike(
      makeTestObject('oops', 'wrong type'),
      referenceObject,
      { checkType: true },
    )).toBe(false);

    // correct type, with type checking
    expect(isObjectLike(
      makeTestObject(34, 'correct type'),
      referenceObject,
      { checkType: true },
    )).toBe(true);
  });

  it('optionally allows extra properties', () => {
    const hasExtraPropD = { a: 'string', b: true, c: 12, d: 'something' };
    const template = { a: 'string', b: true, c: 12 };

    expect(isObjectLike(hasExtraPropD, template, { allowExtraProps: true }))
      .toBeTrue();
  });
});

describe('isObjectWithExpectedProps', () => {
  const testObject = { foo: 42 };

  it('throws if arrayOfProps is not an array of strings', () => {
    expect(() => isObjectWithExpectedProps(testObject, 12)).toThrow();
    expect(() => isObjectWithExpectedProps(testObject, ['allNeedToBeStrings', 12])).toThrow();
    expect(() => isObjectWithExpectedProps(testObject, ['notHere'])).not.toThrow();
  });

  it('returns false if the "object" is not an object', () => {
    expect(isObjectWithExpectedProps(12, ['foo'])).toBe(false);
    expect(isObjectWithExpectedProps(null, ['foo'])).toBe(false);
  });

  it('returns true if the props are there', () => {
    expect(isObjectWithExpectedProps(testObject, ['foo'])).toBe(true);
  });

  it('returns false if the props are not there', () => {
    expect(isObjectWithExpectedProps(testObject, ['notThere'])).toBe(false);
    expect(isObjectWithExpectedProps(testObject, ['foo', 'notThere'])).toBe(false);
  });
});

describe('isPopulatedArray', () => {
  it('tests for array and length', () => {
    expect(isPopulatedArray(12)).toBeFalse();
    expect(isPopulatedArray([])).toBeFalse();
    expect(isPopulatedArray(['finally'])).toBeTrue();
  });
});

describe('isPopulatedObject', () => {
  it('returns false for non-object (non {}) input', () => {
    expect(isPopulatedObject(12)).toBe(false);
    expect(isPopulatedObject(null)).toBe(false);
  });

  it('returns false for {} with no keys and no symbols', () => {
    expect(isPopulatedObject({})).toBe(false);
  });

  it('returns true for {} with keys', () => {
    const onlyHasAmethod = {
      returnA() {
        return 'a';
      },
    };

    expect(isPopulatedObject({ hi: 'there' })).toBe(true);

    expect(isPopulatedObject(onlyHasAmethod)).toBe(true);
  });

  it('returns true for {} with symbols', () => {
    const testSymbol = Symbol('testSymbol');
    // eslint-disable-next-line prefer-const
    let x = {};

    x[testSymbol] = 'hi';

    expect(isPopulatedObject(x)).toBe(true);
  });
});

describe('isPopulatedString', () => {
  it('tests for string and length', () => {
    expect(isPopulatedString(12)).toBeFalse();
    expect(isPopulatedString('')).toBeFalse();
    expect(isPopulatedString('yes!')).toBeTrue();
  });
});

describe('isZeroLength', () => {
  it('tests for a length property equal to zero', () => {
    // no length property
    expect(isZeroLength(12)).toBeFalse();
    expect(isZeroLength({})).toBeFalse();
    expect(isZeroLength({ hasOtherPropeties: true })).toBeFalse();

    expect(isZeroLength('a string!')).toBeFalse();
    expect(isZeroLength(['an array'])).toBeFalse();

    expect(isZeroLength('')).toBeTrue();
    expect(isZeroLength([])).toBeTrue();
  });
});


spec/index.strict.spec.js
/* eslint no-undef:"off" */

const {
  isDate,
  isInteger,
  isObjectLike,
} = require('../index').strict;

// #region jasmine setup
const origTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL;

const revertJasmineTimeout = function revertJasmineTimeout() {
  jasmine.DEFAULT_TIMEOUT_INTERVAL = origTimeout;
};

const setJasmineTimeout = function setJasmineTimeout(milliseconds) {
  jasmine.DEFAULT_TIMEOUT_INTERVAL = milliseconds;
};

// you can set more options than are shown here: see https://jasmine.github.io/api/edge/Reporter.html
// tutorial: https://jasmine.github.io/tutorials/custom_reporter
const myReporter = {
  jasmineStarted: function jasmineStarted(suiteInfo, done) {
    // optional setup goes here
    setJasmineTimeout(10000);
    done();
  },
  jasmineDone: function jasmineDone(suiteInfo, done) {
    console.log(`Tests ended ${new Date().toLocaleString()}`);
    revertJasmineTimeout();
    done();
  },
};

jasmine.getEnv().addReporter(myReporter);
// #endregion jasmine setup

describe('strict mode', () => {
  it('does not change the behavior of tests that do not reference it', () => {
    expect(isDate(1)).toBeFalse();
    expect(isDate(new Date())).toBeTrue();

    expect(isInteger(1)).toBeTrue();
    expect(isInteger(1.009234323423)).toBeFalse();
  });
});

describe('isObjectLike', () => {
  const referenceObject = {
    age: 23,
    name: 'Bob',
    speak: (x) => console.log(x),
  };

  const makeTestObject = function makeTestObject(age = 3, name = 'Charlie') {
    const speak = (x) => console.log(x);
    return { age, name, speak };
  };

  it('sets checkType to true by default', () => {
    expect(isObjectLike(makeTestObject('no type checking', 'Alice'), referenceObject)).toBeFalse();
    expect(isObjectLike(makeTestObject(99, 'Xavier'), referenceObject)).toBeTrue();
  });

  it('returns false if there are more props in the testObject than are in the referenceObject', () => {
    const itHasMoreProps = makeTestObject(72, 'Jans');
    itHasMoreProps.someAdditionalProp = true;
    expect(isObjectLike(itHasMoreProps, referenceObject)).toBe(false);
  });

  it('optionally checks types', () => {
    // wrong type, no type checking
    expect(isObjectLike(
      makeTestObject('oops', 'wrong type'),
      referenceObject,
      { checkType: false },
    )).toBe(true);

    // wrong type, with type checking
    expect(isObjectLike(
      makeTestObject('oops', 'wrong type'),
      referenceObject,
      { checkType: true },
    )).toBe(false);

    // correct type, with type checking
    expect(isObjectLike(
      makeTestObject(34, 'correct type'),
      referenceObject,
      { checkType: true },
    )).toBe(true);
  });

  it('optionally allows extra properties', () => {
    const hasExtraPropD = { a: 'string', b: true, c: 12, d: 'something' };
    const template = { a: 'string', b: true, c: 12 };

    expect(isObjectLike(hasExtraPropD, template, { allowExtraProps: true, debug: true }))
      .toBeTrue();
  });
});


index.js
const main = function main(args = { strict: false }) {
  const checkType = function checkType(x, expected) {
    // eslint-disable-next-line valid-typeof
    return typeof x === expected;
  };

  // #region simple "is" functions

  const isArray = function isArray(x) {
    return Array.isArray(x);
  };

  const isBigInt = function isBigInt(x) {
    return checkType(x, 'bigint');
  };

  const isBoolean = function isBoolean(x) {
    return checkType(x, 'boolean');
  };

  const isDate = function isDate(x) {
    return x instanceof Date;
  };

  const isFunction = function isFunction(x) {
    return checkType(x, 'function');
  };

  const isNull = function isNull(x) {
    return x === null;
  };

  /**
   * @description Checks for a number.  Returns false for NaN.
   * @param {number} x The number to check.
   */
  const isNumber = function isNumber(x) {
    return checkType(x, 'number') &&
      !(Number.isNaN(x));
  };

  /**
   * @description Checks for an object (e.g. { message: 'hi' }).  This is a basic check
   *    that relies on typeof, so perhaps a better name for it is 'is not some other primitive'.
   * @param {object} x The object to test.
   * @returns {boolean} boolean
   * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof
   */
  const isObject = function isObject(x) {
    return checkType(x, 'object') && !(isNull(x));
  };

  const isString = function isString(x) {
    return checkType(x, 'string');
  };

  const isSymbol = function isSymbol(x) {
    return checkType(x, 'symbol');
  };

  const isUndefined = function isUndefined(x) {
    return x === undefined;
  };

  // #endregion simple "is" functions

  // #region more-complex functions

  /**
   * @description Checks for a date greater than (after) someDate.
   *   isDateAfter is an alias of isDateGreaterThan.
   * @param {number} x The number to test.
   * @param {number} someNumber The number to compare to.
   * @returns {boolean} boolean
   */
  const isDateGreaterThan = function isDateGreaterThan(x, someDate) {
    if (!(isDate(someDate))) {
      throw new Error('Expected the second agument to be a date');
    }

    return isDate(x) && (x > someDate);
  };

  /**
   * @description Checks for a date less than (before) someNumber.
   *   isDateBefore is an alias of isDateLessThan.
   * @param {number} x The number to test.
   * @param {number} someNumber The number to compare to.
   * @returns {boolean} boolean
   */
  const isDateLessThan = function isDateLessThan(x, someDate) {
    if (!(isDate(someDate))) {
      throw new Error('Expected the second agument to be a date');
    }

    return isDate(x) && (x < someDate);
  };

  /**
   * @description Checks for a floating point number (a number with a nonzero decimal).
   * @param {number} x The number to test.
   * @returns {boolean} boolean
  */
  const isFloat = function isFloat(x) {
    return isNumber(x) && (x !== Math.floor(x));
  };

  /**
   * @description Checks for an integer.  Allows zero-value decimals e.g. 5.0.
   * @param {number} x The number to test.
   * @returns {boolean} boolean
  */
  const isInteger = function isInteger(x) {
    return isNumber(x) && Number.isInteger(x);
  };

  /**
   * @description Checks for an ISO 8601-compliant date string.  Only checks these formats:
   *    // with or without milliseconds, with the +/- hh:mm at the end
   *    2020-01-11T15:03:11-10:00 || 2020-01-11T15:03:11.999-10:00
   *    // with or without milliseconds, with the Z at the end
   *    2020-01-11T15:03:11Z || 2020-01-11T15:03:11.999Z
   * @param {string} x The string to test.
   * @returns {boolean} boolean
  */
  const isIsoDateTimeString = function isIsoDateTimeString(x) {
    const hhmmNoMilliseconds = /^\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d[+-][0-2]\d:[0-5]\d$/;
    const hhmmWithMilliseconds = /^\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d{1,}[+-][0-2]\d:[0-5]\d$/;
    const zNoMilliseconds = /^\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\dZ$/;
    const zWithMilliseconds = /^\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d{1,}Z$/;

    return isString(x) && (
      zWithMilliseconds.test(x) ||
      zNoMilliseconds.test(x) ||
      hhmmWithMilliseconds.test(x) ||
      hhmmNoMilliseconds.test(x)
    );
  };

  const isNullOrUndefined = function isNullOrUndefined(x) {
    return isNull(x) || isUndefined(x);
  };

  /**
   * @description Checks for a number greater than someNumber.
   * @param {number} x The number to test.
   * @param {number} someNumber The number to compare to.
   * @returns {boolean} boolean
   */
  const isNumberGreaterThan = function isNumberGreaterThan(x, someNumber) {
    if (!(isNumber(someNumber))) {
      throw new Error('Expected the second agument to be a number');
    }

    return isNumber(x) && (x > someNumber);
  };

  /**
   * @description Checks for a number less than someNumber.
   * @param {number} x The number to test.
   * @param {number} someNumber The number to compare to.
   * @returns {boolean} boolean
   */
  const isNumberLessThan = function isNumberLessThan(x, someNumber) {
    if (!(isNumber(someNumber))) {
      throw new Error('Expected the second agument to be a number');
    }

    return isNumber(x) && (x < someNumber);
  };

  const isPopulatedObject = function isPopulatedObject(x) {
    return isObject(x) &&
      (Object.keys(x).length > 0 || Object.getOwnPropertySymbols(x).length > 0);
  };

  const isPopulatedArray = function isPopulatedArray(x) {
    return isArray(x) && x.length > 0;
  };

  const isPopulatedString = function isPopulatedString(x) {
    return isString(x) && x.length > 0;
  };

  /**
   * @description Checks for an object with required properties.
   * @param {object} x The object to test.
   * @param {string[]} arrayOfProps An array of strings with the propery names to test for.
   * @returns {boolean} boolean
   */
  const isObjectWithExpectedProps = function isObjectWithExpectedProps(x, arrayOfProps) {
    if (!(isArray(arrayOfProps)) || !(arrayOfProps.every(isString))) {
      throw new Error('Expected arrayOfProps to be an array of strings.');
    }

    // aviod return arrayOfProps.every((el) => !(isUndefined(x[el])));
    // we get false positives if a prop exists but === undefined
    // e.g. { foo: undefined } will return a false negative for foo
    return isObject(x) && arrayOfProps.every((el) => Object.keys(x).includes(el));
  };

  /**
   * @description Checks that an object has the same properties as a reference object.  Optionally
   *   checks to make sure that the properties are of the same type.  Does not check symbols.
   * @param {object} x The object to test.
   * @param {object} template The object you want x to look like.
   * @param {object} [options] Options.
   * @param {boolean} [options.allowExtraProps=false] Return true if the object to test has
   *   properties not in template and other tests pass.
   * @param {boolean} [options.checkType=false] Strict mode changes the default to true.  Return
   *   false if the properties are there but are of different types.
   * @param {boolean} [options.debug=false] Print troubleshooting info to console.
   * @returns {boolean} boolean
  */
  const isObjectLike = function isObjectLike(
    x,
    template,
    options = {},
  ) {
    if (!(isObject(template))) {
      throw new Error('Expected template to be an object.');
    }

    if (!(isObject(options))) {
      throw new Error('Expected options to be an object.');
    }

    // avoid errors in calculating propertiesCheckPasses
    if (!(isObject(x))) {
      return false;
    }

    // #region set options
    const copyOfOptions = { ...options };
    const defaults = {
      allowExtraProps: false,
      checkType: args.strict,
      debug: false,
    };

    for (let index = 0; index < Object.keys(defaults).length; index++) {
      const currentKey = Object.keys(defaults)[index];

      if (copyOfOptions[currentKey] === undefined) {
        copyOfOptions[currentKey] = defaults[currentKey];
      }
    }
    // #endregion set options

    if (copyOfOptions.debug) {
      console.log('options is:');
      console.log(copyOfOptions);
    }

    const objectPropsAreSameType = function objectPropsAreSameType() {
      return Object.keys(template)
        .every((el) => {
          if (copyOfOptions.debug && (typeof x[el] !== typeof template[el])) {
            console.log(`Property type mismatch detected for property ${el}`);
            console.log(`Got ${typeof x[el]}, expected ${typeof template[el]}`);
          }

          return typeof x[el] === typeof template[el];
        });
    };

    const propertiesCheckPasses = copyOfOptions.allowExtraProps ?
      true :
      Object.keys(x).length === Object.keys(template).length;

    const typeCheckPasses = copyOfOptions.checkType ?
      objectPropsAreSameType() :
      true;

    if (copyOfOptions.debug) {
      console.log(`x is ${JSON.stringify(x, null, 2)}`);
      console.log(`template is ${JSON.stringify(template, null, 2)}`);
      console.log(`isObjectWithExpectedProps? ${isObjectWithExpectedProps(x, Object.keys(template))}`);
      if (isObject(x)) {
        console.log(`key length the same? ${Object.keys(x).length === Object.keys(template).length}`);
      }
      console.log(`objectPropsAreSameType? ${objectPropsAreSameType()}`);
    }

    return isObjectWithExpectedProps(x, Object.keys(template)) &&
      propertiesCheckPasses &&
      typeCheckPasses;
  };

  /**
   * @description Tests for a length property of 0.  Returns false if the length property
   *    doesn't exist.
   * @param {array} somePropName illuminatingDescription.
   * @returns {boolean} boolean
  */
  const isZeroLength = function isZeroLength(x) {
    return !(isNullOrUndefined(x.length)) && x.length === 0;
  };
  // #endregion more-complex functions

  return {
    isArray,
    isBigInt,
    isBoolean,
    isDate,
    isDateAfter: isDateGreaterThan,
    isDateBefore: isDateLessThan,
    isDateGreaterThan,
    isDateLessThan,
    isFunction,
    isFloat,
    isInteger,
    isIsoDateTimeString,
    isNull,
    isNullOrUndefined,
    isNumber,
    isNumberGreaterThan,
    isNumberLessThan,
    isObject,
    isObjectLike,
    isObjectWithExpectedProps,
    isPopulatedArray,
    isPopulatedObject,
    isPopulatedString,
    isString,
    isSymbol,
    isUndefined,
    isZeroLength,
  };
};

module.exports = {
  ...main(),
  strict: { ...main({ strict: true }) },
};
